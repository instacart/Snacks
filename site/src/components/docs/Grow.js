import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

Example:

<Playground>
{`
    initialState = { isOpen: false };
    <div>
      <Button onClick={() => setState({ isOpen: !state.isOpen })} style={{ marginBottom: 10 }}>Toggle</Button>
      <div style={{ display: 'flex' }}>
        <Grow in={state.isOpen}>
          <div style={{padding: 20, display: 'flex', alignItems:'center', background: '#43B02A', flexDirection: 'column', width: 250, border: '1px solid black' }}>
            <LoadingBox shape='circle' />
            <LoadingBox shape='line' />
            <LoadingBox shape='line' />
            <LoadingBox shape='line' />
          </div>
        </Grow>
        <div>
          Content right
        </div>
      </div>
      <div style={{ marginTop: 10 }}>
        Content below
      </div>
    </div>

`}
</Playground>

)
