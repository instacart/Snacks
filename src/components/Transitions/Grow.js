import React, { PureComponent } from 'react'
import Transition from 'react-transition-group/Transition'
import PropTypes from 'prop-types'

const TIMEOUT = 0
const TRASITION_TIME = 200
const TIMING_FUNCTION = 'ease-in-out'
const START_SCALE = 0
const END_SCALE = 1
const START_MAX_SIZE = 0
const END_MAX_SIZE = 1500

const SCALE_DEFAULT = {
  start: 0,
  end: 1,
}
const MAX_SIZE = {
  start: 0,
  end: 1500,
}

const AXIS = 'y'

class Grow extends PureComponent {
  static propTypes = {
    /**
     * A convenience prop that enables or disabled appear animations for
     * all children. Note that specifying this will override any defaults
     * set on individual children Transitions.
     */
    appear: PropTypes.bool,

    /** Axis that is animated */
    axis: PropTypes.oneOf(['x', 'y']),

    children: PropTypes.node,

    /** Show the component; triggers the enter or exit states */
    in: PropTypes.bool,

    /**
     * Settings for max-height and max-width during animation (this is what animates the element's height/width).
     *
     * Default start: 0
     * Default end: 1500 (If content's height is larger than 1500, pass the content's height here )
     */
    maxSize: PropTypes.shape({
      start: PropTypes.number,
      end: PropTypes.number,
    }),

    /**
     * Settings for starting and ending transformation to scale.
     *
     * Default start: 0
     * Default end: 1
     */
    scale: PropTypes.shape({
      start: PropTypes.number,
      end: PropTypes.number,
    }),

    /** Optional style overrides. */
    style: PropTypes.object,

    /** Delay in milliseconds until animation start. */
    timeout: PropTypes.number,

    /** Name of the transition-timing-function CSS property. */
    timingFunction: PropTypes.oneOf(['ease', 'ease-in', 'ease-out', 'ease-in-out', 'linear']),

    /** Time of animation in milliseconds. */
    transitionTime: PropTypes.number,
  }

  static defaultProps = {
    timeout: TIMEOUT,
    transitionTime: TRASITION_TIME,
    timingFunction: TIMING_FUNCTION,
    scale: SCALE_DEFAULT,
    maxSize: MAX_SIZE,
    axis: AXIS,
    style: {},
    in: true,
    appear: true,
  }

  get scale() {
    const { start = START_SCALE, end = END_SCALE } = this.props.scale
    return {
      start,
      end,
    }
  }

  get maxSize() {
    const { start = START_MAX_SIZE, end = END_MAX_SIZE } = this.props.maxSize
    return {
      start,
      end,
    }
  }

  get transformAxis() {
    return this.props.axis === 'x' ? 'scaleX' : 'scaleY'
  }

  get transitionStyles() {
    const { start: startScale, end: endScale } = this.scale
    const { start: startMaxSize, end: endMaxSize } = this.maxSize

    return {
      entering: {
        maxWidth: startMaxSize,
        maxHeight: startMaxSize,
        transform: `${this.transformAxis}(${startScale})`,
      },
      entered: {
        maxWidth: endMaxSize,
        maxHeight: endMaxSize,
        transform: `${this.transformAxis}(${endScale})`,
      },
    }
  }

  get initialStyles() {
    const { start: startScale } = this.scale
    const { start: startMaxSize } = this.maxSize
    const { transitionTime, timingFunction } = this.props

    return {
      maxWidth: startMaxSize,
      maxHeight: startMaxSize,
      transform: `${this.transformAxis}(${startScale})`,
      transition: `all ${transitionTime}ms ${timingFunction}`,
    }
  }

  renderChild = state => {
    const { style, children } = this.props
    const styles = {
      ...this.initialStyles,
      ...this.transitionStyles[state],
      ...style,
    }

    return <div style={styles}>{children}</div>
  }

  render() {
    const { in: inProp, timeout, appear } = this.props

    return (
      <Transition in={inProp} appear={appear} timeout={timeout}>
        {this.renderChild}
      </Transition>
    )
  }
}

export default Grow
