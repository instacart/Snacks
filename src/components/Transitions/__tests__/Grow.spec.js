import React from 'react'
import renderer from 'react-test-renderer'
import Grow from '../Grow'

jest.mock('react-dom', () => ({
  findDOMNode: () => ({
    getContext: jest.fn(),
  }),
}))

it('renders Fade in correctly', () => {
  const tree = renderer
    .create(
      <Grow in>
        <h3> Content </h3>
      </Grow>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders Grow correctly with custom props', () => {
  const scale = {
    start: 0.2,
    end: 0.8,
  }
  const maxSize = {
    start: 10,
    end: 500,
  }
  const tree = renderer
    .create(
      <Grow
        in
        axis="y"
        style={{ background: 'red' }}
        transitionTime={500}
        scale={scale}
        maxSize={maxSize}
      >
        <h3> Content </h3>
      </Grow>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
