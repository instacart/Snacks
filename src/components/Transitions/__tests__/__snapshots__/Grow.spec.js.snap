// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Fade in correctly 1`] = `
.emotion-0 {
  max-width: 0;
  max-height: 0;
  -webkit-transform: scaleY(0);
  -ms-transform: scaleY(0);
  transform: scaleY(0);
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
}

<div
  className="emotion-0"
>
  <h3>
     Content 
  </h3>
</div>
`;

exports[`renders Grow correctly with custom props 1`] = `
.emotion-0 {
  max-width: 10px;
  max-height: 10px;
  -webkit-transform: scaleY(0.2);
  -ms-transform: scaleY(0.2);
  transform: scaleY(0.2);
  -webkit-transition: all 500ms ease-in-out;
  transition: all 500ms ease-in-out;
  background: red;
}

<div
  className="emotion-0"
>
  <h3>
     Content 
  </h3>
</div>
`;
